aspnetcore/security/authentication/mfa.md

```diff
@@ -0,0 +1,560 @@
+
+If MFA is required, and the user in ASP.NET Core Identity has MFA enabled, then the login continues. If the user has no MFA enabled, the user is redirected to the custom view *ErrorEnable2FA.cshtml*. Then ASP.NET Core Identity signs the user in.
+
+```csharp
+//
+// POST: /Account/Login
+[HttpPost]
+[AllowAnonymous]
+[ValidateAntiForgeryToken]
+public async Task<IActionResult> Login(LoginInputModel model)
+{
+	var returnUrl = model.ReturnUrl;
+	var context = await _interaction.GetAuthorizationContextAsync(returnUrl);
+	var requires2Fa = context?.AcrValues.Count(t => t.Contains("mfa")) >= 1;
+
+	var user = await _userManager.FindByNameAsync(model.Email);
+	if(user != null && !user.TwoFactorEnabled && requires2Fa)
+	{
+		return RedirectToAction(nameof(ErrorEnable2FA));
+	}
+```
+
+The ExternalLoginCallback works like the local Identity login. The AcrValues property is checked for the "mfa" value and if it is sent, the MFA is forced before the login completes, ie redirected to the ErrorEnable2FA view.
+
+```csharp
+//
+// GET: /Account/ExternalLoginCallback
+[HttpGet]
+[AllowAnonymous]
+public async Task<IActionResult> ExternalLoginCallback(string returnUrl = null, string remoteError = null)
+{
+	var context = await _interaction.GetAuthorizationContextAsync(returnUrl);
+	var requires2Fa = context?.AcrValues.Count(t => t.Contains("mfa")) >= 1;
+
+	if (remoteError != null)
+	{
+		ModelState.AddModelError(string.Empty, _sharedLocalizer["EXTERNAL_PROVIDER_ERROR", remoteError]);
+		return View(nameof(Login));
+	}
+	var info = await _signInManager.GetExternalLoginInfoAsync();
+
+	if (info == null)
+	{
+		return RedirectToAction(nameof(Login));
+	}
+
+	var email = info.Principal.FindFirstValue(ClaimTypes.Email);
+
+	if (!string.IsNullOrEmpty(email))
+	{
+		var user = await _userManager.FindByNameAsync(email);
+		if (user != null && !user.TwoFactorEnabled && requires2Fa)
+		{
+			return RedirectToAction(nameof(ErrorEnable2FA));
+		}
+	}
+
+	// Sign in the user with this external login provider if the user already has a login.
+	var result = await _signInManager
+		.ExternalLoginSignInAsync(
+			info.LoginProvider, 
+			info.ProviderKey, 
+			isPersistent: 
+			false);
+```
+
+If the user is already logged in, the client application still validates the "amr" claim, and can setup the MFA then with a link to the ASP.NET Core Identity view.
+
+![acr_values-1](mfa/_static/acr_values-1.png)
+
+## Force ASP.NET Core OpenID Connect client to require MFA
+
+This example shows how an ASP.NET Core Razor Page app, which uses OpenID Connect to sign in, can require that users have authenticated using MFA.
+
+To validate the MFA requirement, an IAuthorizationRequirement requirement is created. This will be added to the pages using a policy which require MFA.
```

Comment:
```suggestion
To validate the MFA requirement, an `IAuthorizationRequirement` requirement is created. This will be added to the pages using a policy which require MFA.
```